Make Xamarin.Build.Download FIPS Compliant

Redth opened the pull request.

Redth committed:
[xbd] Add support for new `AndroidAarLibrary` item

Xamarin.Android in 15.7 and newer will include the `AndroidAarLibrary` item group.  .aar files added to this group will automatically be compiled into the app, so Xamarin.Build.Download no longer has to embed them into the referenced assembly that needs them as embedded resources.

Redth committed:
[xbd] Move aar item support to property

Also adds the include target to the master list of targets to run

Redth committed:
[xbd] fix logical test (copy paste error)

Redth committed:
[xbd] remove extra quote

Redth committed:
[xbd] Remove temp AvailableItemName declaration

Redth committed:
[xbd] preview release

Redth committed:
[xbd] Rework availableitemname checking

The property group is evaluated before the itemgroups, so we would never detect an availableitemname in the previous case.

This moves the detection straight into the target conditions.

Redth committed:
[xbd] fix proguard config target dependency

We won’t always run _XamarinAndroidBuildAarRestore, so we should instead depend on _XamarinBuildDownload to ensure the .aar files are in place before we try and process them.

Redth committed:
[xbd] Fix name, AvailableItemName not Names

newky2k committed:
Merge branch 'master' into xbd-aar-15_7

Redth committed:
Merge branch 'master' into xbd-aar-15_7

Redth committed:
Switch from MD5/SHA1

To CRC/SHA256

Redth committed:
Switch to xunit

Redth committed:
Use netstandard

Redth committed:
Better 7zip location

Redth committed:
Merge branch 'master' into xbd-aar-15_7

Redth committed:
Remove unnecessary references for netstandard2

Redth committed:
Allow tests to ignore error mesasge patterns

Redth committed:
Helper method for tests to determine windows

Redth committed:
Ignore some failure message patterns

The netstandard msbuild bits throw some warnings we can ignore

Redth committed:
Some tests will never work on windows

Windows Visual Studio has some file locking issues with the old ways we used to rewrite referenced assemblies to include the correct xamarin.build.download'd bits.  They will never work on windows.

Redth committed:
Remove md5/sha1 from tests

Redth committed:
modernize cake

Redth committed:
Add nuget package info

Redth committed:
Fix msbuild prop spec

Redth committed:
Build on mac

Redth committed:
Fix path to xunit test assembly

Redth committed:
Try multiple test paths (local and ci)

Redth committed:
Copy test refs to output dir for running tests

Redth committed:
Use dotnet test runner

Redth committed:
Bump to dotnetcore 3.0

mattleibow committed:
test ci

mattleibow committed:
Instead of bailing out fast, try be useful

mattleibow committed:
Remove the line

mattleibow committed:
Merge branch 'master' into xbd-fips

mattleibow commented on Util/Xamarin.Build.Download/build.cake:
```diff
@@ -1,28 +1,54 @@
+		if (!string.IsNullOrEmpty(SOURCE_BRANCH))
+			c.Properties.Add("RepositoryBranch", new [] { SOURCE_BRANCH });
+		if (!string.IsNullOrEmpty(SOURCE_COMMIT))
+			c.Properties.Add("RepositoryCommit", new [] { SOURCE_COMMIT });
```
Since we may actually want to force this on all builds, we could do this in the template:

https://github.com/xamarin/XamarinComponents/blob/master/.ci/build.yml#L187-L188

mattleibow commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download.Tests/HashTests.cs:
```diff
@@ -0,0 +1,24 @@
+    public class HashTests
+    {
+        [Theory]
+        [InlineData("C:\\path\\to\\file.aar")]
```
This will fail on mac with invalid paths. I think?

Redth replied:
The intent is just to see that i get no funny characters (just numbers, and digits) in the CRC code returned.  I'll add a test case for a mac path too for kicks.

mattleibow commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download.Tests/Helpers/TestsBase.cs:
```diff
@@ -28,8 +27,10 @@ public void SetUp ()
 
-		[OneTimeTearDown]
-		public void TearDown ()
+        public bool IsWindows
```
I know this is just text, but I still hope the files will have either tabs or spaces.

mattleibow commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download/XamarinBuildGetArchivesToDownload.cs:
```diff
@@ -57,8 +57,8 @@ public override bool Execute ()
-					if (!string.IsNullOrEmpty (partialZipDownload.Md5))
-						taskItem.SetMetadata ("Md5", partialZipDownload.Md5);
+					if (!string.IsNullOrEmpty (partialZipDownload.Sha256))
+						taskItem.SetMetadata ("Sha256", partialZipDownload.Sha256);
```
I see all these nice changes, but are we backwards compatible?

Redth replied:
It should just skip over verifying the old md5 hashes since it no longer checks for them and no longer sets the task item, it will just never be checked to be validated.

mattleibow commented:
/azp run

mattleibow committed:
Trying this

mattleibow committed:
keep the directory structure

mattleibow committed:
Set git variables and use trx extension

mattleibow committed:
Go back to master

mattleibow committed:
Merge branch 'master' into xbd-fips

mattleibow committed:
Align whitespaces

mattleibow committed:
Merge branch 'master' into xbd-fips

Redth committed:
Hide

Redth committed:
fix tests

Redth committed:
Update HashTests.cs

Redth committed:
Rename md5 to sha256 in error

Redth committed:
Test for proguard text differently

Redth committed:
Merge remote-tracking branch 'origin/xbd-aar-15_7' into xbd-fips

mattleibow committed:
Merge branch 'master' into xbd-fips

mattleibow commented:
/azp run

mattleibow commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download.Tests/Test.cs:
```diff
@@ -471,7 +487,6 @@ public void TestDisallowUnsafeGetItemsToDownload ()
 					{ "ToFile", "cardview.v7.aar" },
 					{ "RangeStart", "196438127" },
 					{ "RangeEnd", "196460160" },
-					{ "Md5", "b44eb88f7cc621ae616744c6646f5b64" }
```
We are removing a check to make sure that the file is both correct/complete as well as the right file. We need to have a new way to do this. Since maven only has SHA1 and MD5, we are probably going to have to do the hashing ourselves on CI or as part of the binding process.

mattleibow commented on Util/Xamarin.Build.Download/source/Xamarin.Build.Download/Xamarin.Build.Download.csproj:
```diff
@@ -1,94 +1,41 @@
     </None>
+    <None Include="$(TargetPath)" Pack="True" PackagePath="build\Xamarin.Build.Download.dll" />
+    <None Include="$(PkgMono_Cecil)\lib\netstandard2.0\Mono.Cecil.dll" Visible="False" Pack="True" PackagePath="build\Mono.Cecil.dll" />
+    <None Include="$(PkgNewtonsoft_Json)\lib\netstandard2.0\Newtonsoft.Json.dll" Visible="False" Pack="True" PackagePath="build\Newtonsoft.Json.dll" />
```
I like this.

mattleibow commented on manifest.yaml:
```diff
@@ -592,7 +592,8 @@
   TriggerPaths: [ Util/Xamarin.Build.Download ]
-  MacBuildTargets: [ nuget ]
+  MacBuildTargets: [ ci ]
+  WindowsBuildTargets: [ ci ]
```
I like this pattern of adding both the Windows and Mac agents in. Should start showing any cross-platform issues before it even gets to the gate.

mattleibow reviewed (approved):
Looking good. 2 tests seem to be failing.

Redth committed:
Disable some now irrelevant tests

Redth committed:
Merge branch 'xbd-fips' of ssh://github.com/xamarin/XamarinComponents into xbd-fips

Redth committed:
Merge branch 'master' into xbd-fips

Redth committed:
Try for local install of 7z

This should address the idea in https://github.com/xamarin/XamarinComponents/pull/716

Redth committed:
Move aar fixups into downloader

Redth committed:
Remove old android .dll embed tasks

Redth committed:
Fix the flagrant mixture of whitespace

Redth committed:
Remove Sha1

Redth committed:
Open zip properly

Redth committed:
More sha1 removal

Redth committed:
Fix duplicate . in filename

Redth committed:
Fix type/bad value

Redth committed:
Fix conditions for available items

Redth committed:
Cleanup tests

Redth committed:
Update Test.cs

Redth committed:
Old test cleanup

mattleibow reviewed (approved):
LGTM

Redth committed:
Remove old .aar nesting folder fixup

When we used to embed .aar's inside .dll's as a resource, Xamarin.Android expected everything in that .aar (well really .zip) to be nested inside a `library_project_imports` directory, but now that is not necessary.

Redth committed:
Merge branch 'master' into xbd-fips

Redth committed:
Stable version XBD package
